backend/services/user_groups.go

```diff
@@ -0,0 +1,421 @@
+// Copyright 2022, Yahoo Inc.
+// Licensed under the terms of the MIT. See LICENSE file in project root for terms.
+
+package services
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/theparanoids/ashirt-server/backend"
+	"github.com/theparanoids/ashirt-server/backend/database"
+	"github.com/theparanoids/ashirt-server/backend/dtos"
+	"github.com/theparanoids/ashirt-server/backend/models"
+	"github.com/theparanoids/ashirt-server/backend/policy"
+	"github.com/theparanoids/ashirt-server/backend/server/middleware"
+
+	sq "github.com/Masterminds/squirrel"
+)
+
+type CreateUserGroupInput struct {
+	Name      string
+	Slug      string
+	UserSlugs []string
+}
+
+type ModifyUserGroupInput struct {
+	Name          string
+	Slug          string
+	UsersToAdd    []string
+	UsersToRemove []string
+}
+
+type ListUserGroupsForAdminInput struct {
+	UserGroupFilter
+	Pagination
+	IncludeDeleted bool
+}
+
+type ListUserGroupsForOperationInput struct {
+	Pagination
+	UserGroupFilter
+	OperationSlug string
+}
+
+type userGroupAndRole struct {
+	models.UserGroup
+	Role policy.OperationRole `db:"role"`
+}
+
+type ListUserGroupsInput struct {
+	Query          string
+	IncludeDeleted bool
+	OperationSlug  string
+}
+
+func (i ModifyUserGroupInput) validateUserGroupInput() error {
+	if i.Slug == "" {
+		return backend.MissingValueErr("Slug")
+	}
+	return nil
+}
+
+func (i CreateUserGroupInput) validateUserGroupInput() error {
+	if i.Slug == "" {
+		return backend.MissingValueErr("Slug")
+	}
+	if i.Name == "" {
+		return backend.MissingValueErr("Name")
+	}
+	return nil
+}
+
+func AddUsersToGroup(tx *database.Transactable, userSlugs []string, groupID int64) error {
+	if len(userSlugs) > 0 {
+		questionMarks := "("
```

Comment:
This is a style thing, but Go tends to prefer early exits, so you'd normally see:
```go
if len(userSlugs) == 0 {
  return nil
}
```
Doesn't really matter though.

Reply:
noted!